README.md
# Twitchshot
A program that scrapes Twitch.tv clips and compiles them into a video.

Functioning as a content aggregation tool, Twitchshot is aimed towards content creators who want to compile
the most popular clips of videogames from streamers on Twitch.tv.

This program is not in any way affiliated with Twitch or Amazon.


CHANGELOG.txt
# Changelog

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
The date format is YYYY-MM-DD. The version format is X.Y.Z.

## [0.2.0] - 2020-08-11

### Added -
- Flask Integration
- HTML, CSS
- videoEditor.py

### Changed - 
- Nearly everything.

### Notes -
I haven't been maintaining the changelog, as most of my development was local. I'll be working to 
keep better track as of now, and for each merge to this repository I'll make notes in here about it.
Regarding the last note, input validation has been implemented.

## [0.1.2] - 2020-07-22

### Added -
- Imported moviepy 
- Added clip_data value 'filename'

### Changed - 
- Optimized validateClips()
- Reformatted Twitch method names and calls to make more sense.

### Notes - 
There's no input validation for the Twitch parameters. I should have some indication at
least for the filter, because that relies on pre-defined keywords. There also isn't 
custom directory implementation -- I'll get that put in tomorrow.


## [0.1.1] - 2020-07-20

### Added - 
- Implemented method t.generateEmbed()


## [0.1.0] - 2020-07-19 

### Added -
- Implemented clip validation through t.validateClips()
- Implemented clip value ['clip_valid']
- Created t.downloadClips()
- Added t.print() function in class Twitch to print values in a given clip_data dict.

### Changed - 
- Moved testing code into function v.downloadClips in class Video.

### Notes -
The current implementation of validating clips is atrocious from a user standpoint.
All of the clip thumbnails are opened one by one, and they can't be closed as of yet. This will be much
better once the PyQT GUI is implemented, but for now it's pretty terrible.
I think I'll move on to PyQT implementation tomorrow. As of now, the program is capable of a full
retrieval and download of clips given a certain game, so I want a proper GUI before I move on to
concatenation and intro/outro.


tsForms.py
import os

from flask_wtf import FlaskForm
from wtforms import (
                      StringField,
                      SelectField,
                      IntegerField,
                      SubmitField,
                      PasswordField
                    )
from wtforms.validators import DataRequired, NumberRange, Length


class TwitchForm(FlaskForm):

  csrf_token = os.urandom(32)  # CSRF Token. It has to be declared here and referenced in the HTML. 
  game_name = StringField('Game Name', validators=[DataRequired()])
  time_range = SelectField('Time Range', choices=[('LAST_DAY', 'Last Day'), ('LAST_WEEK', 'Last Week'), ('LAST_MONTH', 'Last Month'), ('ALL_TIME', 'All Time')])
  clip_amount = IntegerField('Clip Amount (1-50)', [NumberRange(min=1, max=50, message='Please enter a number between one and fifty.')]) 
  submit = SubmitField('Submit')

# class TestingForm(FlaskForm):

#   csrf_token = os.urandom(32)
#   clip_index = IntegerField('Clip Number', [NumberRange(min=0, message='Please enter a valid index.')])
#   submit = SubmitField('Submit')

class ChoiceForm(FlaskForm):

  csrf_token = os.urandom(32)
  approve = SubmitField('Approve')
  reject = SubmitField('Reject')
  submit = SubmitField('Submit')

class RenderForm(FlaskForm):

  csrf_token = os.urandom(32)
  name = StringField('Final Video Name', validators=[DataRequired(), Length(min=1, max=100, message='Too long of a filename. Does it really need to be that long?')])
  submit = SubmitField('Start Rendering')

twitchshot.py
import os
from threading import Thread
import concurrent.futures

from flask import Flask, render_template, request, redirect, url_for, session
from flask_wtf.csrf import CSRFProtect

from twitchClips import Twitch, validateGame
from videoEditor import Video
from tsForms import TwitchForm, ChoiceForm, RenderForm

app = Flask(__name__)
csrf = CSRFProtect(app)
app.config.from_pyfile('tsConfig.py')


@app.route('/', methods=['POST', 'GET'])
def index():

    session.clear()
    session['finished'] = False
    form = TwitchForm()

    # The form is properly filled out.
    if form.validate_on_submit():
        game = form.game_name.data
        time = form.time_range.data
        amount = str(form.clip_amount.data)

        # checks if game name is valid
        if validateGame(game):
            return redirect(url_for('clips', game=game, time=time, amount=amount, clip_index=0))
        else:
            return redirect(url_for('error'))

    else:
        return render_template('tsIndex.html', form=form)


@app.route('/clips/<game>/<time>/<amount>/<clip_index>', methods=['POST', 'GET'])
def clips(game, time, amount, clip_index):

    # This means it is the first time the user reaches the page.
    if 'clips' not in session:

        twt = Twitch(game, int(amount), time)

        with concurrent.futures.ThreadPoolExecutor() as executor:
            clips = list(executor.map(twt.retrieveClipData, twt.retrieveLinks()))

        session['clips'] = clips

        return redirect(url_for('clips', game=game, time=time, amount=amount, clip_index=str(0)))

    # One of the three choice buttons was pressed.
    if request.method == 'POST':
        
        clip_index = int(clip_index)

        if(request.values.has_key('approve')):
           return redirect(validate('true', game, time, amount, clip_index))

        if(request.values.has_key('reject')):
            return redirect(validate('false', game, time, amount, clip_index))
        
        if(request.values.has_key('submit')):
            approved_clips = {}
            
            # Creates a key:value pair of valid-filename:download-link.
            for clip in session['clips']:
                if clip['valid'] == 'true':
                    approved_clips[clip['filename']] = clip['dl']
            
            session['downloads'] = approved_clips
            return redirect(url_for('render'))

    form = ChoiceForm()
    return render_template('tsClips.html', game=game, clips=session['clips'], display_clip=session['clips'][int(clip_index)], current_index=str(clip_index), form=form)


@app.route('/render', methods=['POST', 'GET'])
def render():
    
    form = RenderForm()

    # get confirmation from user to start -- therefore a post request to this page
    if form.validate_on_submit():

        v = Video(form.name.data)
        # v.compositeVideo(session['downloads'])
        session['download_link'] = v.uploadVideo()
        session['finished'] = True

        return redirect(url_for('finished'))

    else:
        
        return render_template('tsRender.html', form=form)

@app.route('/finished')
def finished():

    return render_template('tsFinished.html', download=session['download_link'])

@app.route('/about/')
def about():

    return '<h1>oh hey this is an about page</h1>'

@app.route('/error/')
def error():
    return f'<h1>ya blew it</h1>'


# Handles approval/rejection of clips.
def validate(boolean, game, time, amount, clip_index):

    temp_storage = session['clips']  # Sets the sesssion to a temporary variable to ensure valid changes.

    temp_storage[clip_index]['valid'] = boolean  # Sets the clip to the given 'boolean'.

    if((clip_index + 1) < (len(temp_storage))):  # Moves the user to the next clip if there is one, restarts the index if there isn't.
        session['clips'] = temp_storage
        return (url_for('clips', game=game, time=time, amount=amount, clip_index=clip_index+1))
    else:
        session['clips'] = temp_storage
        return (url_for('clips', game=game, time=time, amount=amount, clip_index=0))


if __name__ == "__main__":
    app.run()


videoEditor.py
# Logan Halverson
# Personal, videoEditor.py
# 05 August 2020

import os
import requests
import time
import shutil
import random

from slugify import slugify
import moviepy.editor as mpy

class Video():

    def __init__(self, render_name):
        
        self.video_filenames = []
        self.video_file_clips = []
        self.render_name = render_name
        self.id = str(int(999999999999 * random.random()))
        
        # Makes a randomized folder name and works within it for each user.
        
        if not os.path.basename(os.getcwd()) == 'video':
           os.chdir(os.path.dirname(os.path.abspath(__file__)) + '\\video')

        os.mkdir(self.id)
        os.chdir(self.id)

    # Downloads the video from the links, saves as valid filename.
    def downloadVideo(self, links):

        for fname, dlink in links.items():

            filename = fname + '.mp4'
            self.video_filenames.append(filename)
            with open(filename, 'wb') as f:
                video = requests.get(dlink)
                f.write(video.content)
                self.video_file_clips.append(mpy.VideoFileClip(filename))

    def compositeVideo(self, links):

        # Downloads the video from the links, saves as valid filename.
        for fname, dlink in links.items():
            
            filename = fname + '.mp4'
            self.video_filenames.append(filename)

            with open(filename, 'wb') as f:
                video = requests.get(dlink)
                f.write(video.content)
                self.video_file_clips.append(mpy.VideoFileClip(filename))

        # Concatenates the videos.
        final_clip = mpy.concatenate_videoclips(self.video_file_clips, method="compose")
        self.render_name = slugify(self.render_name)
        self.render_name += '.mp4'
        final_clip.write_videofile(self.render_name)


    def uploadVideo(self):

        # DELETE USER INSTANCE OF VIDEO FOLDER
        os.chdir(os.path.dirname(os.path.abspath(__file__)) + '\\video')

        # # This has to be done not during the render process, but some time after with
        # # a cleanup function. It can't modify files in use.
        # shutil.rmtree(self.id)
        
        # DROPBOX/GOOGLE DRIVE upload rendered video, return download link to FLASK app
        return "you're mom gay"

# links = {
#     'superman64' : 'downloadlink1',
#     'mariogowoo' : 'downloadlink3',
#     'heeohooaa' : 'downloadlink2'
# }


properTwitchClipBot.py
# Logan Halverson
# Personal, properTwitchClipBot.py
# 19 July 2020

import os
import requests
from slugify import slugify
from PIL import Image
import moviepy.editor as mpy

'''
    os : used to handle directories.
    requests : used to make HTTP requests and download videos.
    slugify : used to ensure valid filenames from clip titles.
    image : used to handle previewing clips.
    mpy : used to work with video files
'''

class Twitch():


    # Initializes on a specified game, the time range, and how many clips to return.
    def __init__(self, clip_game, clip_amount=20, clip_filter='LAST_WEEK'):
        self.clip_game = clip_game
        self.clip_amount = clip_amount
        self.clip_filter = clip_filter
        
        self.session = requests.session()
        self.session.headers = {
            "Client-ID": "kimne78kx3ncx6brgo4mv6wki5h1ko"  # !!!!! - I don't know if this specific ID works for people that aren't me.
        }
    

    # Given the class variables, creates a specific request and retrieves the top clip links.
    def retrieveClipLinks(self):

        clip_links = []

        payload = [{"operationName":"ClipsCards__Game","variables":{"gameName":self.clip_game,"limit":self.clip_amount,"criteria":{"languages":[],"filter":self.clip_filter}},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"0d8d0eba9fc7ef77de54a7d933998e21ad7a1274c867ec565ac14ffdce77b1f9"}}}]

        r = self.session.post("https://gql.twitch.tv/gql",json=payload)

        for clip in r.json()[0]["data"]["game"]["clips"]["edges"]:
            clip_links.append(clip['node']['url'])

        return clip_links


    # Given the link to a clip, retrieves all useful data from it.
    def retrieveClipData(self, link):

        # The slug is the auto-generated clip name, found in the url. The HTTP request
        # works off of that, so it is retrieved from the url and stored as String slug.
        slug = link[24:]
        
        # This is the headers the HTTP request is sent with.
        payload = [ {"operationName":"ClipsView","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"4480c1dcc2494a17bb6ef64b94a5213a956afb8a45fe314c66b0d04079a93a8f"}}},{"operationName":"ClipsViewCount","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"00209f168e946123d3b911544a57be26391306685e6cae80edf75cdcf55bd979"}}},{"operationName":"ClipsTitle","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"f6cca7f2fdfbfc2cecea0c88452500dae569191e58a265f97711f8f2a838f5b4"}}},{"operationName":"VideoAccessToken_Clip","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"9bfcc0177bffc730bd5a5a89005869d2773480cf1738c592143b5173634b7d15"}}}, {"operationName":"WatchLivePrompt","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"d65226c25ec2335f7550351c7041f4080a2531f4e330375fef45c7a00e4f4016"}}}]

        r = self.session.post('https://gql.twitch.tv/gql', json=payload)
        data = r.json()

        '''
        CLIP_DATA VALUES:

            clip_url : A url linking to the clip itself on twitch.tv
            clip_dl : A direct video download link for the highest quality available
            clip_thumbnail : A image link to the thumbnail of the clip
            clip_game : The game the clip is from
            clip_author : The channel the clip is from
            clip_views : The views the clip has
            clip_title : The title of the clip, made by the user who clipped it
            clip_filename : A valid OS filename generated from the clip_title.

        '''

        clip_data = {
            'clip_url' : link,
            'clip_dl' : data[3]['data']['clip']['videoQualities'][0]['sourceURL'],
            'clip_thumbnail' : data[4]['data']['clip']['thumbnailURL'],
            'clip_game' : data[0]['data']['clip']['game']['name'],
            'clip_author' : data[0]['data']['clip']['broadcaster']['displayName'],
            'clip_views' : data[1]['data']['clip']['viewCount'],
            'clip_title' : data[2]['data']['clip']['title'],
        }

        clip_data['clip_filename'] = slugify(clip_data['clip_title'])

        return clip_data


    # Packages all retrieved clips into a list and returns it.
    def retrieveClips(self):
        clip_links = self.retrieveClipLinks()
        clips = []

        for link in clip_links:
            clips.append(self.retrieveClipData(link))

        return clips


    # Overriding print() to print all data about a clip given a clip dict.
    def print(self, clip_data):
        print('URL : ' + clip_data['clip_url'])
        print('DOWNLOAD : ' + clip_data['clip_dl'])
        print('THUMBNAIL : ' + clip_data['clip_thumbnail'])
        print('GAME : ' + clip_data['clip_game'])
        print('TITLE : ' + clip_data['clip_title'])
        print('FILENAME : ' + clip_data['clip_filename'])
        print('AUTHOR : ' + clip_data['clip_author'])
        print('VIEWS : ' + str(clip_data['clip_views']))
        print()


    # This method goes through a list of clips and shows the user the thumbnail of each.
    # It then removes any clip not approved by the user from the list.
    # This method is also **godawful** from a user persepective. This will be changed once we
    # reach the GUI implementation stage, but as of now its just honestly atrocious.
    def validateClips(self, clips):
        
        print("We'll be reviewing the clips by looking at their thumbnail.")
        print("If the clip looks valid, type OK.")
        valid_clips = []

        for clip in clips:

            filename = clip['clip_filename'] + '.jpg'
            target = os.path.join(os.getcwd(), filename)

            # print('Filename : {0}\nTarget : {1}'.format(filename, target))

            with open(target, 'wb') as f:
                r = requests.get(clip['clip_thumbnail'])
                f.write(r.content)
                
            im = Image.open(filename)
            im.show()

            answer = input('Approve clip? ')
            if answer.lower() == 'ok': 
                valid_clips.append(clip)

            os.remove(filename)

        return valid_clips

    # Given a clip, generates the embed link that will be used on a website.
    def generateEmbed(self, clip, parent, height=378, width=620):  
        slug = clip['clip_url'][24:]
        template = '<iframe src="https://clips.twitch.tv/embed?clip={0}&parent={1}" frameborder="0" allowfullscreen="true" scrolling="no" height="{2}" width="{3}"></iframe>'.format(slug, parent, height, width)
        return template


class Video():

    def downloadClips(self, clips):

        dirname = os.path.dirname(__file__)

        for clip in clips:
            directory = (os.path.join(dirname, clip['clip_filename'] + ".mp4"))

            with open(directory, 'wb') as f:
                r = requests.get(clip['clip_dl'])
                f.write(r.content)
                print('Successfully saved ' + clip['clip_filename'])


t = Twitch('valorant', 5, 'LAST_WEEK') 
v = Video()

# All necessary steps so far are taken using these three methods.
clips = t.retrieveClips()
#clips = t.validateClips(clips)
v.downloadClips(clips)



twitchClips.py
# Logan Halverson
# Personal, properTwitchClipBot.py
# 19 July 2020

import os
import requests
import time
from slugify import slugify

'''
    os : used to handle directories.
    requests : used to make HTTP requests and download videos.
    slugify : used to ensure valid filenames from clip titles.
'''


class Twitch():


    # Initializes on a specified game, the time range, and how many clips to return.
    def __init__(self, clip_game, clip_amount=20, clip_filter='LAST_WEEK'):
        self.clip_game = clip_game
        self.clip_amount = clip_amount
        self.clip_filter = clip_filter  # LAST_DAY, LAST_WEEK, LAST_MONTH, ALL_TIME
        self.clip_links = []
        
        self.session = requests.session()
        self.session.headers = {
            "Client-ID": "kimne78kx3ncx6brgo4mv6wki5h1ko",  # !!!!! - I don't know if this specific ID works for people that aren't me.
        }
    

    # Prints the most popular games on Twitch by viewer count.
    def printTopGames(self, limit=10):
        
        payload = [{"operationName":"BrowsePage_AllDirectories","variables":{"limit":limit,"options":{"recommendationsContext":{"platform":"web"},"requestID":"JIRA-VXP-2397","sort":"VIEWER_COUNT","tags":[]}},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"78957de9388098820e222c88ec14e85aaf6cf844adf44c8319c545c75fd63203"}}}]

        r = self.session.post('https://gql.twitch.tv/gql', json=payload)
        data = r.json()[0]['data']['directoriesWithTags']['edges']

        print('The Top {0} Games'.format(limit))

        for x in range(limit):
            name = data[x]['node']['name']
            viewers = data[x]['node']['viewersCount']
            print('{0}. {1} with {2} viewers.'.format(x+1, name, viewers))

    # Given the class variables, creates a specific request and retrieves the top clip links.
    def retrieveLinks(self, cursor=""):

        if cursor == "":
            payload = [{"operationName":"ClipsCards__Game","variables":{"gameName":self.clip_game,"limit":self.clip_amount,"criteria":{"languages":[],"filter":self.clip_filter}},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"0d8d0eba9fc7ef77de54a7d933998e21ad7a1274c867ec565ac14ffdce77b1f9"}}}]
        else:
            payload = [{"operationName":"ClipsCards__Game","variables":{"gameName":self.clip_game,"limit":(self.clip_amount - len(self.clip_links)),"criteria":{"languages":[],"filter":self.clip_filter}, "cursor":cursor},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"0d8d0eba9fc7ef77de54a7d933998e21ad7a1274c867ec565ac14ffdce77b1f9"}}}]

        r = self.session.post("https://gql.twitch.tv/gql",json=payload)

        try:
            for clip in r.json()[0]["data"]["game"]["clips"]["edges"]:
                self.clip_links.append(clip['node']['url'])

            assert len(self.clip_links) == self.clip_amount

            return self.clip_links

        except TypeError:
            print("Reached TypeError Exception!")
            return []
        
        except AssertionError:
            print("Didn't return the amount of clips requested! Moving with cursor...")
            cursor = (r.json()[0]["data"]["game"]["clips"]["edges"])[-1]["cursor"]
            print(f"Remaining Clips to Retrieve : {self.clip_amount - len(self.clip_links)}")
            print(f"Cursor Value : {cursor}")
            self.clip_links = self.retrieveLinks(cursor)
            return self.clip_links

            
    def retrieveData(self, clip_links):
        clips = []
        for link in clip_links:
            clips.append(self.retrieveClipData(link))
        return clips

    # Given the link to a clip, retrieves all useful data from it.
    def retrieveClipData(self, link):

        # The slug is the auto-generated clip name, found in the url. The HTTP request
        # works off of that, so it is retrieved from the url and stored as String slug.
        slug = link[24:]
        
        # This is the headers the HTTP request is sent with.
        payload = [ {"operationName":"ClipsView","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"4480c1dcc2494a17bb6ef64b94a5213a956afb8a45fe314c66b0d04079a93a8f"}}},{"operationName":"ClipsViewCount","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"00209f168e946123d3b911544a57be26391306685e6cae80edf75cdcf55bd979"}}},{"operationName":"ClipsTitle","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"f6cca7f2fdfbfc2cecea0c88452500dae569191e58a265f97711f8f2a838f5b4"}}},{"operationName":"VideoAccessToken_Clip","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"9bfcc0177bffc730bd5a5a89005869d2773480cf1738c592143b5173634b7d15"}}}, {"operationName":"WatchLivePrompt","variables":{"slug":slug},"extensions":{"persistedQuery":{"version":1,"sha256Hash":"d65226c25ec2335f7550351c7041f4080a2531f4e330375fef45c7a00e4f4016"}}}]

        r = self.session.post('https://gql.twitch.tv/gql', json=payload)
        data = r.json()

        '''
            url : A url linking to the clip itself on twitch.tv
            dl : A direct video download link for the highest quality available
            thumbnail : A image link to the thumbnail of the clip
            game : The game the clip is from
            author : The channel the clip is from
            views : The views the clip has
            title : The title of the clip, made by the user who clipped it
            filename : A valid OS filename generated from the clip_title.
            embed : An HTML embed link made from method t.generateEmbed().
        '''

        clip_data = {
            'url' : link,
            'dl' : data[3]['data']['clip']['videoQualities'][0]['sourceURL'],
            'thumbnail' : data[4]['data']['clip']['thumbnailURL'],
            'game' : data[0]['data']['clip']['game']['name'],
            'author' : data[0]['data']['clip']['broadcaster']['displayName'],
            'views' : data[1]['data']['clip']['viewCount'],
            'title' : data[2]['data']['clip']['title'],
        }

        clip_data['filename'] = slugify(clip_data['title'])
        clip_data['embed'] = self.generateEmbed(clip_data['url'], str('127.0.0.1'))
        clip_data['valid'] = 'true'

        return clip_data


    # Given a clip, generates the embed link that will be used on a website.
    def generateEmbed(self, clip_url, parent):  
        slug = clip_url[24:]
        template = "https://clips.twitch.tv/embed?clip={0}&parent={1}".format(slug, parent)
        return template

def validateGame(game):

    headers = {
        'Client-ID' : 'kimne78kx3ncx6brgo4mv6wki5h1ko',
        'Accept' : 'application/vnd.twitchtv.v5+json'
    }

    r = requests.get('https://api.twitch.tv/kraken/clips/top', headers=headers, params={'game' : game})

    if r.json()['clips'] == []:
        return False  # Invalid game.
    else:
        return True